osu.Game.Tests/Visual/Gameplay/TestSceneClicksPerSecond.cs

```diff
@@ -0,0 +1,339 @@
+
+            for (double i = 1; i <= 2; i += 0.25)
+            {
+                changeRate(i);
+                double rate = i;
+                AddAssert($"KPS approx. = {i}", () => MathHelper.ApproximatelyEquivalent(calculator!.Value, 10 * rate, 0.5));
+            }
+
+            for (double i = 1; i >= 0.5; i -= 0.25)
+            {
+                changeRate(i);
+                double rate = i;
+                AddAssert($"KPS approx. = {i}", () => MathHelper.ApproximatelyEquivalent(calculator!.Value, 10 * rate, 0.5));
+            }
+        }
+
+        [Test]
+        public void TestInputsDiscardedOnRewind()
+        {
+            createCalculator();
+            startClock();
+
+            AddStep("Create consistent KPS inputs", () => addInputs(generateConsistentKps(10)));
+            seek(1000);
+
+            AddAssert("KPS = 10", () => calculator!.Value == 10);
+
+            AddStep("Create delayed inputs", () => addInputs(generateConsistentKps(10, 50)));
+            seek(1000);
+            AddAssert("KPS didn't changed", () => calculator!.Value == 10);
+        }
+
+        private void seekAllClocks(double time)
+        {
+            gameplayClockContainer?.Seek(time);
+            manualClock!.CurrentTime = time;
+        }
+
+        protected override Ruleset CreateRuleset() => new OsuRuleset();
+
+        #region Quick steps methods
+
+        private void createCalculator()
+        {
+            AddStep("create calculator", () =>
+            {
+                Debug.Assert(dependencyContainer?.Dependencies.Get(typeof(DrawableRuleset)) is DrawableRuleset);
+                dependencyContainer!.Children = new Drawable[]
+                {
+                    calculator = new ClicksPerSecondCalculator(),
+                    new DependencyProvidingContainer
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                        CachedDependencies = new (Type, object)[] { (typeof(ClicksPerSecondCalculator), calculator) },
+                        Child = new ClicksPerSecondCounter // For visual debugging, has no real purpose in the tests
+                        {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                            Scale = new Vector2(5),
+                        }
+                    }
+                };
+                calculator!.Listener = listener = new ManualInputListener(calculator!);
+            });
+        }
+
+        private void seek(double time) => AddStep($"Seek clocks to {time}ms", () => seekAllClocks(time));
+
+        private void changeRate(double rate) => AddStep($"Change rate to x{rate}", () => manualClock!.Rate = rate);
+
+        private void advanceForwards(double time) =>
+            AddStep($"Advance clocks {time} seconds forward.", () =>
+            {
+                gameplayClockContainer!.Seek(gameplayClockContainer.CurrentTime + time * manualClock!.Rate);
```

Comment:
We don't use `!`. make the field non-null and use `= null!;` (search code base for thousands of examples).